### ISSUE TITLE
Blocking main thread due to observer

### ISSUE THREAD

**Issue by shannonhochkins** [measuredco/auto-frame-component#1]
When using the puck editor, my application does have a lot of styles, it's using emotion so there's going to be lots of individual style tags.

I get `[Violation] 'message' handler took 17110ms` in my console, and when i click on the `index.js:139` it takes me to the observer within this package. It's taking 17 seconds to unblock the main thread and then im able to use the puck editor, has this been reported before?

I believe if you expose the frame reference element, I can create an emotion cache provider to tell emotion to add the styles to the iframe instead of this observer having to copy all the styles from the parent document, by default, emotion is adding them outside the frame which is causing this long delay.

![Screenshot 2024-04-16 at 9 54 04 am](https://github.com/measuredco/auto-frame-component/assets/5105002/c687648c-7f9e-4a03-a2e0-d589e56f21ae)

**Comment by chrisvxd** [comment]
Hi @shannonhochkins! I believe this might be related to [#407](https://github.com/measuredco/puck/issues/407) (long thread, but they're using emotion) and should be resolved in auto-frame-component@0.1.5.

I've been attempting to patch this - can you verify whether Puck [0.14.2-canary.e585f20](https://www.npmjs.com/package/@measured/puck/v/0.14.2-canary.e585f20) canary fixes it? 

The authors of #407 are reporting a new issue (iframe not appearing at all), but I'm unable to reproduce.

**Comment by shannonhochkins** [comment]
Thanks legend, I'll check this out, although the correct way to process this is to have a cache provider with emotion and have it render directly into the iframe so the observer doesn't need to move/watch anything

I managed to (almost) solve this using a hack to retrieve the iframe however there's still some styles that aren't created within the iframe which I suspect is due to the rendering nature of nextjs!

Managed to reduce the load times from 17s to 3s, and will continue to debug with your changes

**Comment by chrisvxd** [comment]
Interesting. The challenge  I've been facing is coming up with a general purpose solution to avoid too much fiddling in user-land. Perhaps there are more domain-specific configs we could provide when using iframes, so interested to see what you figure out.

For what it's worth, I've just published another fix under 0.14.2-canary.717ccb9 that addresses the new issue reported in #407. There's still one outstanding issue, but I believe this is a broader dnd issue than anything style related.

**Comment by shannonhochkins** [comment]
Without sounding like a dick, I've spent a stupid amount of time on performance improvements on applications :D Do you mind if i suggest some changes?

In the last section where you append the mirror elements to the head, try using a document fragment, it may be quicker as  it allows you to build a sub-tree in memory without causing page reflow or repaint for each node added. Only when the complete fragment is appended to the DOM does it cause a reflow and repaint, which is more efficient than multiple individual operations.

```
const styleFragments = document.createDocumentFragment();
        mirrors.forEach(mirror => {
            if (mirror) {
                styleFragments.appendChild(mirror);
            }
        });
        doc.head.appendChild(styleFragments);
```

There might a particular reason why you're using a setTimeout with 0ms, however with setTimeout using it the way you are it could lead to multiple separate updates within a single frame, or even across frames


`setTimeout(fn, 0)` effectively places the callback at the end of the current event queue, which could lead to execution at any point in time, potentially right before or during a rendering process. This can cause frame drops if the browser was ready to paint the frame but must wait for the callback to complete, which might not be synchronized with the browser’s refresh rate.

You could simply try wrapping it in a RAF but i'm not sure of the intent behind the deferring logic!



```
requestAnimationFrame(() => {
    addedElements.forEach(el => addEl(el));  // Batch process all added elements
    removedElements.forEach(el => removeEl(el));  // Batch process all removed elements
});
```

**Comment by shannonhochkins** [comment]
@chrisvxd Just tried the canary version, can confirm it's no rendering without blocking the main thread at all! Nice work mate!

**Comment by chrisvxd** [comment]
@shannonhochkins

No problem re suggesting changes, I'm all ears!

Sounds like the performance suggestions aren't valid for the latest version, but feel free to open an subsequent PRs with any further performance improvements if you can see any quick wins.
